core/log-parser/log-readers/sftptail.js

```diff
@@ -0,0 +1,164 @@
+import crypto from 'crypto';
+import EventEmitter from 'events';
+import fs from 'fs';
+import { readFile } from 'fs/promises';
+import path from 'path';
+import SftpClient from 'ssh2-sftp-client';
+
+export default class SFTPTail extends EventEmitter {
+  constructor(options = {}) {
+    super();
+    this.options = options;
+
+    // Setup basic-ftp client.
+    this.client = new SftpClient();
+
+    // Setup logger.
+    if (typeof this.options.log === 'function') {
+      this.log = this.options.log;
+    } else if (this.options.log) {
+      this.log = console.log;
+    } else {
+      this.log = () => {};
+    }
+
+    // Setup internal properties.
+    this.filePath = options.path;
+    this.lastByteReceived = 0;
+
+    this.fetchLoopActive = false;
+    this.fetchLoopPromise = null;
+  }
+
+  async watch() {
+    // Setup temp file.
+    this.tmpFilePath = path.join(
+      process.cwd(),
+      crypto
+        .createHash('md5')
+        .update(`${this.options.host}:${this.options.port}:${this.filePath}`)
+        .digest('hex') + '.tmp'
+    );
+
+    // Connect.
+    await this.connect();
+
+    // Start fetch loop.
+    this.log('Starting fetch loop...');
+    this.fetchLoopActive = true;
+    this.fetchLoopPromise = this.fetchLoop();
+  }
+
+  async unwatch() {
+    this.log('Stopping fetch loop...');
+    this.fetchLoopActive = false;
+    await this.fetchLoopPromise;
+  }
+
+  async fetchLoop() {
+    while (this.fetchLoopActive) {
+      try {
+        // Store the start time of the loop.
+        const fetchStartTime = Date.now();
+
+        // Reconnect the FTP client in case it has disconnected.
+        const connection = await this.connect();
+
+        // Get the size of the file on the FTP server.
+        this.log(`Fetching size of file...${this.filePath}`);
+        const fileSize = await connection.stat(this.filePath).size;
+        this.log(`File size is ${fileSize}.`);
+
+        // If the file size has not changed then skip this loop iteration.
+        if (fileSize === this.lastByteReceived) {
+          this.log('File has not changed.');
+          await this.sleep(this.options.fetchInterval);
+        }
+
+        // If the file has not been tailed before or it has been decreased in size download the last
+        // few bytes.
+        if (this.lastByteReceived === null || this.lastByteReceived > fileSize) {
```

Comment:
you never do a check about the file size when it starts to track and this causes script to start reading from first line to last line at once on start